src/Core/Registration/ColorMapOptimization.cpp

```diff
@@ -0,0 +1,683 @@
+#include <Core/Geometry/RGBDImage.h>
+#include <Core/Geometry/TriangleMesh.h>
+#include <Core/Camera/PinholeCameraTrajectory.h>
+#include <IO/ClassIO/TriangleMeshIO.h>
+#include <Core/Utility/Eigen.h>
+
+namespace three {
+
+namespace {
+
+class ImageWarpingField {
+
+public:
+    ImageWarpingField (int width, int height, int number_of_vertical_anchors) {
+        InitializeWarpingFields(width, height, number_of_vertical_anchors);
+    }
+
+    void InitializeWarpingFields(int width, int height,
+            int number_of_vertical_anchors) {
+        anchor_h_ = number_of_vertical_anchors;
+        anchor_step_ = (double)height / (anchor_h_ - 1);
+        anchor_w_ = std::ceil((double)width / anchor_step_) + 1;
+        flow_ = Eigen::VectorXd(anchor_w_ * anchor_h_ * 2);
+        for (int i = 0; i <= (anchor_w_ - 1); i++) {
+            for (int j = 0; j <= (anchor_h_ - 1); j++) {
+                int baseidx = (i + j * anchor_w_) * 2;
+                flow_(baseidx) = i * anchor_step_;
+                flow_(baseidx + 1) = j * anchor_step_;
+            }
+        }
+    }
+
+    Eigen::Vector2d QueryFlow(int i, int j) const {
+        int baseidx = (i + j * anchor_w_) * 2;
+        // exceptional case: quried anchor index is out of pre-defined space
+        if (baseidx < 0 || baseidx > anchor_w_ * anchor_h_ * 2)
+            return Eigen::Vector2d(0.0, 0.0);
+        else
+            return Eigen::Vector2d(flow_(baseidx), flow_(baseidx + 1));
+    }
+
+    Eigen::Vector2d GetImageWarpingField(double u, double v) const {
+        int i = (int)(u / anchor_step_);
+        int j = (int)(v / anchor_step_);
+        double p = (u - i * anchor_step_) / anchor_step_;
+        double q = (v - j * anchor_step_) / anchor_step_;
+        return (1 - p) * (1 - q) * QueryFlow(i, j)
+            + (1 - p) * (q)* QueryFlow(i, j + 1)
+            + (p)* (1 - q) * QueryFlow(i + 1, j)
+            + (p)* (q)* QueryFlow(i + 1, j + 1);
+    }
+
+public:
+    Eigen::VectorXd flow_;
+    int anchor_w_, anchor_h_;
+    double anchor_step_;
+};
+
+const double IMAGE_BOUNDARY_MARGIN = 10;
+
+inline std::tuple<float, float, float> Project3DPointAndGetUVDepth(
+        const Eigen::Vector3d X,
+        const PinholeCameraTrajectory& camera, int camid)
+{
+    std::pair<double, double> f = camera.intrinsic_.GetFocalLength();
+    std::pair<double, double> p = camera.intrinsic_.GetPrincipalPoint();
+    Eigen::Vector4d Vt = camera.extrinsic_[camid] *
+            Eigen::Vector4d(X(0), X(1), X(2), 1);
+    float u = (Vt(0) * f.first) / Vt(2) + p.first;
+    float v = (Vt(1) * f.second) / Vt(2) + p.second;
+    float z = Vt(2);
+    return std::make_tuple(u, v, z);
+}
+
+std::tuple<std::vector<std::vector<int>>, std::vector<std::vector<int>>>
+        MakeVertexAndImageVisibility(const TriangleMesh& mesh,
+        const std::vector<RGBDImage>& images_rgbd,
+        const PinholeCameraTrajectory& camera,
+        const ColorMapOptmizationOption& option)
```

Comment:
I think back when I made visibility test, I mask all vertices around depth discontinuity as invisible. This is because at depth discontinuity, the depth image becomes noisy, the measurement is not accurate in terms of both depth and x-y value. The projection also gets unstable. This happens frequently when you do a 360 scan of an object. So I chose to mask visible vertices only when:
1.  projected depth is close to the depth in depth image
2. the point on depth image should have a normal close to the viewing ray

Can you try a sequence which is a 360 scan and see if you see any boundary ghosting? If so, the pixels around depth discontinuity need to be eroded.

Reply:
This also applies when averaging pixel values. Only points with a good normal orientation (not too far away from the viewing ray) gets averaged.